oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/MetadataRegistry.java

```diff
@@ -0,0 +1,433 @@
+     */
+    private List<TagMetadata> parseExtraTagSpecs(String tags, Schema.SchemaBuilder builder) {
+        if (StringUtil.isEmpty(tags)) {
+            return Collections.emptyList();
+        }
+        String[] tagsArray = tags.split(",");
+        if (tagsArray.length == 0) {
+            return Collections.emptyList();
+        }
+        List<TagMetadata> extraTagMetadataList = new ArrayList<>();
+        for (final String tagName : tagsArray) {
+            builder.spec(tagName, new ColumnSpec(ColumnType.TAG, String.class));
+            extraTagMetadataList.add(new TagMetadata(parseIndexRule(tagName, null),
+                    TagFamilySpec.TagSpec.newStringTag(tagName)));
+        }
+        return extraTagMetadataList;
+    }
+
+    /**
+     * Parse TagSpec from {@link ModelColumn}
+     *
+     * @param modelColumn the column in the model to be parsed
+     * @return a typed tag spec
+     */
+    @Nullable
+    private TagFamilySpec.TagSpec parseTagSpec(ModelColumn modelColumn) {
+        final Class<?> clazz = modelColumn.getType();
+        final String colName = modelColumn.getColumnName().getStorageName();
+        if (String.class.equals(clazz) || DataTable.class.equals(clazz) || JsonObject.class.equals(clazz)) {
+            return TagFamilySpec.TagSpec.newStringTag(colName);
+        } else if (int.class.equals(clazz) || long.class.equals(clazz)) {
+            return TagFamilySpec.TagSpec.newIntTag(colName);
+        } else if (byte[].class.equals(clazz)) {
+            return TagFamilySpec.TagSpec.newBinaryTag(colName);
+        } else if (Layer.class.equals(clazz)) {
+            return TagFamilySpec.TagSpec.newIntTag(colName);
+        } else if (double.class.equals(clazz) || Double.class.equals(clazz)) {
+            // serialize double as binary
+            return TagFamilySpec.TagSpec.newBinaryTag(colName);
+        } else if (IntList.class.isAssignableFrom(clazz)) {
+            return TagFamilySpec.TagSpec.newIntArrayTag(colName);
+        } else { // handle exceptions
+            // TODO: we skip all tags with type of List<String>
+            if ("tags".equals(colName)) {
+                return null;
+            }
+            throw new IllegalStateException("type " + modelColumn.getType().toString() + " is not supported");
+        }
+    }
+
+    public PartialMetadata parseMetadata(Model model) {
+        if (model.isRecord()) {
+            String group = "stream-default";
+            if (model.isSuperDataset()) {
+                // for superDataset, we should use separate group
+                group = "stream-" + model.getName();
+            }
+            return new PartialMetadata(group, model.getName(), Kind.STREAM);
+        }
+        return new PartialMetadata("measure-default", model.getName(), Kind.MEASURE);
+    }
+
+    @RequiredArgsConstructor
+    @Data
+    public static class PartialMetadata {
```

Comment:
Why do you name this `PartialMetadata`? As metadata is not completed, only partial?
This seems a little strange.

Reply:
Fixed